### ISSUE TITLE
[FEATURE]: Load the reward model in Stage 3 training script of ChatGPT

### ISSUE THREAD

**Issue by snowyday** [hpcaitech/ColossalAI#3011]
### Describe the feature

I would like to propose an update to the code in [`train_prompts.py`](https://github.com/hpcaitech/ColossalAI/blob/main/applications/ChatGPT/examples/train_prompts.py) for Stage 3 of the project. Currently, the reward model is copied from the critic model defined using the same pre-trained model specified by `args.pretrain`, without loading the reward model created in Stage 2.

https://github.com/hpcaitech/ColossalAI/blob/82503a96f20c2e61016af78445f9d1697ac07dd5/applications/ChatGPT/examples/train_prompts.py#L44

To address this, I suggest replacing Line 44 of "train_prompts.py" with the following code:

```
if args.model == 'bloom':
    reward_model = BLOOMRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
elif args.model == 'opt':
    reward_model = OPTRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
elif args.model == 'gpt2':
    reward_model = GPTRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
else:
    raise ValueError(f'Unsupported model "{args.model}"')

strategy.load_model(reward_model, args.rmmodel, strict=False) # NOTE: args.rmmodel = "rm_checkpoint.pt"
```
This code properly loads the reward model created in Stage 2 for use in Stage 3, based on the model specified by the user.

Thank you for your consideration of this request (I'd be willing to do some initial work on this request myself).

**Comment by FrankLeeeee** [comment]
Hi @Fazziekey  and @ht-zhou , can you help to answer this question?

**Comment by taishiciR** [comment]
"This code properly loads the reward model " -- hi @snowyday 
does the model behave diffrently if using the function you preposed earlier "reward_model.load_state_dict(torch.load(args.rmmodel))"

**Comment by Fazziekey** [comment]
> ### Describe the feature
> I would like to propose an update to the code in [`train_prompts.py`](https://github.com/hpcaitech/ColossalAI/blob/main/applications/ChatGPT/examples/train_prompts.py) for Stage 3 of the project. Currently, the reward model is copied from the critic model defined using the same pre-trained model specified by `args.pretrain`, without loading the reward model created in Stage 2.
> 
> https://github.com/hpcaitech/ColossalAI/blob/82503a96f20c2e61016af78445f9d1697ac07dd5/applications/ChatGPT/examples/train_prompts.py#L44
> 
> To address this, I suggest replacing Line 44 of "train_prompts.py" with the following code:
> 
> ```
> if args.model == 'bloom':
>     reward_model = BLOOMRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
> elif args.model == 'opt':
>     reward_model = OPTRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
> elif args.model == 'gpt2':
>     reward_model = GPTRM(pretrained=args.pretrain, lora_rank=args.lora_rank).cuda()
> else:
>     raise ValueError(f'Unsupported model "{args.model}"')
> 
> strategy.load_model(reward_model, args.rmmodel, strict=False) # NOTE: args.rmmodel = "rm_checkpoint.pt"
> ```
> 
> This code properly loads the reward model created in Stage 2 for use in Stage 3, based on the model specified by the user.
> 
> Thank you for your consideration of this request (I'd be willing to do some initial work on this request myself).

Yes, your proposal is very right, we also want to make the all RLHF process become a pipeline, and the reward model should be loaded from previous process, this issue will be open util we finish the pipeline

**Comment by snowyday** [comment]
@FrankLeeeee 
Thank you, that's great! I'm looking forward to the RLHF pipeline.

**Comment by snowyday** [comment]
Hi, @taishiciR  
All the models are handled by the strategy, so I utilized the load function from the strategy class.

https://github.com/hpcaitech/ColossalAI/blob/ea0b52c12ee7598fe126ed0d8b0557f7e8a0e999/applications/ChatGPT/examples/train_prompts.py#L75

However, as you mentioned, I think the functionality is the same since all strategies are currently using `torch.load` for loading.

https://github.com/hpcaitech/ColossalAI/blob/ea0b52c12ee7598fe126ed0d8b0557f7e8a0e999/applications/ChatGPT/chatgpt/trainer/strategies/naive.py#L45

**Comment by binmakeswell** [comment]
We have updated a lot. Please check the latest code. This issue was closed due to inactivity. Thanks.
